src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -473,6 +503,77 @@ public void RemoveAdditionalFile(string fullPath)
 
         #endregion
 
+        #region Non Source File Addition/Removal
+        public void AddDynamicSourceFile(string dynamicFilePath, ImmutableArray<string> folders)
+        {
+            var extension = FileNameUtilities.GetExtension(dynamicFilePath)?.TrimStart('.');
+            if (extension?.Length == 0)
+            {
+                return;
+            }
+
+            foreach (var provider in _dynamicFileInfoProviders)
+            {
+                // skip unrelated providers
+                if (!provider.Metadata.Extensions.Any(e => string.Equals(e, extension, StringComparison.OrdinalIgnoreCase)))
+                {
+                    continue;
+                }
+
+                // don't get confused by _filePath and filePath
+                // VisualStudioProject._filePath points to csproj/vbproj of the project
+                // and the parameter filePath points to dynamic file such as cshtml and etc
+                // 
+                // also, provider is free-threaded. so find to call Wait rather than JTF
+                var fileInfo = provider.Value.GetDynamicFileInfoAsync(
+                    projectId: Id, projectFilePath: _filePath, filePath: dynamicFilePath, CancellationToken.None).WaitAndGetResult_CanCallOnBackground(CancellationToken.None);
+
+                if (fileInfo == null)
+                {
+                    continue;
+                }
+
+                fileInfo = FixUpDynamicFileInfo(fileInfo, dynamicFilePath);
+
+                // remember map between original dynamic file path to DynamicFileInfo.FilePath
+                _dynamicFilePathMaps.Add(dynamicFilePath, fileInfo.FilePath);
+                _sourceFiles.AddDynamicFile(provider.Value, fileInfo, folders);
+                return;
+            }
+        }
+
+        private DynamicFileInfo FixUpDynamicFileInfo(DynamicFileInfo fileInfo, string filePath)
+        {
+            // we might change contract and just throw here. but for now, we keep existing contract where one can return null for DynamicFileInfo.FilePath
+            if (string.IsNullOrEmpty(fileInfo.FilePath))
+            {
+                return new DynamicFileInfo(filePath, fileInfo.SourceCodeKind, fileInfo.TextLoader, fileInfo.DocumentServiceProvider);
+            }
+
+            return fileInfo;
+        }
+
+        public void RemoveDynamicSourceFile(string dynamicFilePath)
+        {
+            var provider = _sourceFiles.RemoveDynamicFile(GetDynamicFileInfoPath(dynamicFilePath));
+
+            // provider is free-threaded. so find to call Wait rather than JTF
+            provider.RemoveDynamicFileInfoAsync(
+                projectId: Id, projectFilePath: _filePath, filePath: dynamicFilePath, CancellationToken.None).Wait(CancellationToken.None);
+        }
```

Comment:
i'm confused by the ownership/order of operations.  If we're being asked to remove a dynamic file... why are we telling a provider to remove hte file?  If htey're the things that provides dynamic files, wouldn't they be the ones telling *us* to remove the file?

Reply:
project system is the one who tells everyone whether file is added or removed. razor never talk to project system directly. we are the middle man. we need to propagate the info to the razor.